qiskit/transpiler/passes/synthesis/plugin.py

```diff
@@ -0,0 +1,359 @@
+
+        This is expected to return a dictionary where the key is a string
+        basis and the value is a list of gate names that the basis works in.
+        If the synthesis method doesn't support multiple basis this should
+        return ``None``.
+
+        If a dictionary is returned by this method the run kwargs will be
+        passed a parameter ``matched_basis`` which contains a list of the
+        basis strings which match the target basis gate set for the
+        transpilation.
+        """
+        pass
+
+    @abc.abstractmethod
+    def run(self, unitary, **options):
+        """Run synthesis for the given unitary matrix
+
+        Args:
+            unitary (numpy.ndarray): The unitary matrix to synthesize to a
+                :class:`~qiskit.dagcircuit.DAGCircuit` object
+            options: The optional kwargs that are passed based on the output
+                of :meth:`supports_basis_gates`, :meth:`supports_coupling_map`,
+                and :meth:`supports_approximation_degree`. If
+                :meth:`supports_coupling_map` returns ``True`` a kwarg
+                ``coupling_map`` will be passed either containing ``None`` (if
+                there is no coupling map) or a
+                :class:`~qiskit.transpiler.CouplingMap` object. If
+                :meth:`supports_basis_gates` returns ``True`` then a kwarg
+                ``basis_gates`` will the list of basis gate names will be
+                passed. Finally if :meth:`supports_approximation_degree`
+                returns ``True`` a kwarg ``approximation_degree`` containing
+                a float for the approximation value will be passed.
+
+        Returns:
+            DAGCircuit: The dag circuit representation of the unitary. Alternatively,
+                you can return a tuple of the form ``(dag, wires)`` where ``dag`` is the dag
+                circuit representation of the circuit representation of the unitary
+                and ``wires`` is the mapping wires to use for
+                :meth:`qiskit.dagcircuit.DAGCircuit.substitute_node_with_dag`.
+        """
+        pass
+
+
+class UnitarySynthesisPluginManager:
+    """Unitary Synthesis plugin manager class
+
+    This class tracks the installed plugins, it has a single property,
+    ``ext_plugins`` which contains a list of stevedore plugin objects.
+    """
+
+    def __init__(self):
+        self.ext_plugins = stevedore.ExtensionManager(
+            "qiskit.unitary_synthesis", invoke_on_load=True, propagate_map_exceptions=True
+        )
+
+
+def unitary_synthesis_plugin_names():
+    """Return a list of installed unitary synthesis plugin names
+
+    Returns:
+        list: A list of the installed unitary synthesis plugin names. The
+            plugin names are valid values for the
+            :func:`~qiskit.compiler.transpile` kwarg
+            ``unitary_synthesis_method``.
+    """
+    plugins = UnitarySynthesisPluginManager()
+    return plugins.ext_plugins.names()
```

Comment:
I just tested this locally and it's a bit more involved than I though. If I make this a shared global there ends up being a import time cycle (although not a fatal one) that blocks stevedore from finding the default plugin since `plugin.py` loads `unitary_synthesis.py` as part of the plugin loading and `unitary_synthesis.py` depends on `plugin.py` for the abstract class definition (and the global). This causes the plugin loading for `default` to fail and not be included in the plugin list. I think for now it'll just be easier to leave it as is even if it's a little weird. I'll leave a comment here about it.

Reply:
Oh ugh, yeah, that's pretty gross.  Best leave it as it is.

Reply:
I added a comment here in: https://github.com/Qiskit/qiskit-terra/pull/6124/commits/ba6a4a0ca67e6a43c85f13d0975509b7f39c3272